### PR TITLE
Port to Go

### PR DESCRIPTION
**Pull request by subpop**
Rewrite the worker in Go, refactoring the IPC to use D-Bus instead of gRPC.

* `meson` is used to build the executable and install the various data files required (D-Bus policies, systemd units, etc.)
* Adds a top-level compile target named `srpm` to create an SRPM to simplify building SRPMs. See `./dist/srpm` for details.
* The Ansible collections previously bundled and shipped as part of the downstream RPM in dist-git have been moved upstream; they are now installed by `meson` during `meson install`.
* The various Python modules required to run `ansible-runner` are installed by `meson` during `meson install`.
* `ansible-runner` is invoked as a subprocess (see `ansible.go`). `ansible-runner` events are read by inotify file monitoring and piped back into the worker. Each time an event file is closed, the worker reads the file and transmits it. This is different behavior than the old version of the worker. That worker would collect events in a dictionary and post them after a timeout. This "response_interval" option will be added in a follow-up PR.

It is my hope that just this rewrite alone addresses some of the long-standing challenges with receiving data back from the worker in a timely manner. However, I would like to see it tested **a lot** so that we can continue to improve it.

I have not tested this at all against a "live" playbook-dispatcher service.

This commit is intense. It drops all the Python code and adds the Go project in one switch to maintain CI builds. When reviewing this PR, it's probably easiest to checkout the branch and examine it as a standalone result rather than trying to read the diffs.

Card ID: CCT-640

### CODE REVIEW

File: CONTRIBUTING.md
Comment (m-horky) [inline CONTRIBUTING.md:45]: For some reason, this isn't working on my system:
```
usage: meson [-h] {setup,configure,dist,install,introspect,init,test,wrap,subprojects,rewrite,compile,devenv,env2mfile,help} ...
meson: error: unrecognized arguments: -build_srpm=True

```

File: main.go
Comment (m-horky) [inline main.go:20]: Please, let's not lock ourselves into this `Go`ism of string flags with a single dash. This is a public API we have to keep, let's use `--flag`-style configuration using some library.

**Review by m-horky — CHANGES_REQUESTED** [review]
I've quickly looked at it, without trying to connect it to yggdrasil bits. _Over all_ it looks sane, I don't have major notes about the architecture.

File: main.go
Comment (subpop) [reply main.go:20]: For user-facing software like `yggctl`, `yggd, and `rhc`, I completely agree with you. I'd use a more fully-featured library. For software like this, that isn't meant to be run frequently by users manually, I'm less concerned with the fluency of the API. Using the standard library's `flag` package means fewer third-party package dependencies. I find that preferable in cases like this where the user experience isn't as high a priority.

Comment (subpop) [reply main.go:20]: And for what it's worth, double-dash flags are parsed as equivalent to the single-dash flags.

```console
$ /usr/libexec/rhc-worker-playbook --version
0.2.0
```

File: CONTRIBUTING.md
Comment (subpop) [reply CONTRIBUTING.md:45]: Whoops. This is a syntax error in my instructions.

File: worker.go
Comment (jirihnidek) [inline worker.go:63]: Unhandled error.

File: python/Cython-0.29.37-py2.py3-none-any.whl
Comment (jirihnidek) [inline python/Cython-0.29.37-py2.py3-none-any.whl:1]: Could we verify that this file comes from right source somehow?

**Review by jirihnidek — CHANGES_REQUESTED** [review]
I have several comments and requests.

I tried to test it. There were many obstacles. I was able to install worker to the testing system using `sudo meson install -C ./builddir`, but starting worker was terminated with following error:

```
[root@fedora github]# systemctl start com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service
Job for com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service failed because the control process exited with error code.
See "systemctl status com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service" and "journalctl -xeu com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service" for details.
[root@fedora github]# systemctl status com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service
× com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service - rhc-worker-playbook worker service
     Loaded: loaded (/usr/lib/systemd/system/com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service; disabled; preset: disabled)
    Drop-In: /usr/lib/systemd/system/service.d
             └─10-timeout-abort.conf
     Active: failed (Result: exit-code) since Tue 2024-08-06 16:25:39 CEST; 18s ago
       Docs: https://github.com/RedHatInsights/rhc-worker-playbook
    Process: 6332 ExecStart=/usr/local/libexec/rhc-worker-playbook (code=exited, status=1/FAILURE)
   Main PID: 6332 (code=exited, status=1/FAILURE)
        CPU: 6ms

Aug 06 16:25:39 fedora systemd[1]: Starting com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service - rhc-worker-playbook worker service...
Aug 06 16:25:39 fedora rhc-worker-playbook[6332]: 2024/08/06 16:25:39 ERROR cannot connect error="cannot request name on bus: Request to own name refused by policy"
Aug 06 16:25:39 fedora systemd[1]: com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service: Main process exited, code=exited, status=1/FAILURE
Aug 06 16:25:39 fedora systemd[1]: com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service: Failed with result 'exit-code'.
Aug 06 16:25:39 fedora systemd[1]: Failed to start com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service - rhc-worker-playbook worker service.
```

File: ansible/ansible-posix-1.5.4.tar.gz
Comment (subpop) [reply ansible/ansible-posix-1.5.4.tar.gz:1]: I'll add a README explaining their origin; that is a good idea. These two files are the Ansible galaxy collections that rhc-worker-playbook includes as part of its Ansible functionality. Old versions of rhc-worker-playbook pulled these files as Sources in the spec file. I chose to move that responsibility upstream so that all the build logic is upstream in a single location.

I include the tarballs as part of project so that it can be built offline. Maybe Fedora-based build farms build packages with networking disabled to ensure the packages don't download files during build time. They're committed as objects in git; is that not sufficient hashing? Under what circumstances will these files be redownloaded from external sources, if not to update them to newer versions?

File: data/dbus/com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.conf.in
Comment (subpop) [reply data/dbus/com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.conf.in:5]: Yes. This worker expects to run as root. I thought so too, but this policy also allows `yggd` (running as non-root user) to send messages to `com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook` (running as root user), so we still have to specify permission for the yggdrasil-worker group to communicate with the bus-names owned by a root process.

File: data/systemd/com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service.in
Comment (subpop) [reply data/systemd/com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service.in:8]: I believe this makes the process run with `yggdrasil-worker` as the group instead of defaulting to the `root` group.

File: dist/srpm/rhc-worker-playbook.spec.in
Comment (subpop) [reply dist/srpm/rhc-worker-playbook.spec.in:1]: I have been testing it with yggdrasil 0.4.2 installed on my system, then rebuilding the SRPM in mock. Instructions similar to:

```
rm -rf builddir
meson setup -Dbuild_srpm=True builddir
meson compile srpm -C builddir
mock --rebuild ./builddir/dist/srpm/*.src.rpm
```

And then installing the RPM built by mock. I forget the output path, but it's something like `/var/lib/mock/fedora-40-x86_64/result/rhc-worker-playbook*.x86_64.rpm`.

Another way to built it is to use packit:

```console
packit build in-mock
```

Either way, you'll probably want to use a mock config that includes our `@yggdrasil/latest` COPR repository, since that repository provides build dependencies like insights-client on Fedora. You can generate such a config using `copr-cli`:

```console
copr-cli mock-config @yggdrasil/latest fedora-40-x86_64 > ~/.config/mock/fedora-40-x86_64-yggdrasil.cfg
```

Then you can use it when building with mock: `mock --root ~/.config/mock/fedora-40-x86_64-yggdrasil.cfg` or with packit: `packit build in-mock --root ~/.config/mock/fedora-40-x86_64-yggdrasil.cfg`.

File: internal/ansible/ansible.go
Comment (subpop) [reply internal/ansible/ansible.go:117]: I don't feel like it's necessary to debug-print this, since we're already
debug-printing each event. Maybe a happy-path info message like "sent event"
after the event gets sent to the channel?

File: internal/ansible/ansible.go
Comment (subpop) [reply internal/ansible/ansible.go:131]: That's okay right? We don't care about other file suffixes, so they can be
safely ignored.

File: internal/ansible/ansible.go
Comment (subpop) [reply internal/ansible/ansible.go:132]: Again, I don't see why it matters that we log about every file we *don't* care about.

File: internal/exec/exec.go
Comment (subpop) [reply internal/exec/exec.go:2]: Perhaps it is. I never considered it complicated enough to search for a third-party package that adds a little process management around the standard library's exec package. This is neatly encapsulated into a package so that it *can* act as a standalone package, if you want to publish one. I didn't feel like it adds enough features to warrant publishing as a standalone execution/process manager package.

File: python/Cython-0.29.37-py2.py3-none-any.whl
Comment (subpop) [reply python/Cython-0.29.37-py2.py3-none-any.whl:1]: Is this something pip could manage for us? Can it be given a file with package names and run a checksum on them for us?

Comment (subpop) [reply python/Cython-0.29.37-py2.py3-none-any.whl:1]: I refactored the Python wheel installation process. I added a classic SHA256SUM file that includes the SHA256 hashes of the included wheels and tarballs. This is checked during build time using `sha256sum --check`. Additionally, I modified the pip installation command to install from a `requirements.txt` file that includes `--hash` flags. This makes pip verify the hashes of the build sources before installing them. I feel like leaving them both, taking a belt & suspenders approach here. The pip hash method was very opaque and difficult to discern what's happening and how.

File: main.go
Comment (subpop) [reply main.go]: I refactored `main` to use `urfave/cli` instead of `ff`.

File: ansible/ansible-posix-1.5.4.tar.gz
Comment (jirihnidek) [reply ansible/ansible-posix-1.5.4.tar.gz:1]: I can see that Fedora and EPEL repositories contains these two collection as RPMs `ansible-collection-ansible-posix.noarch` and `ansible-collection-community-general.noarch`. I guess that we cannot install these collections from RPM, because it is not part of official RHEL repositories, right?

Do we have any mechanism to automatically check for security updates of these two collections?

File: CONTRIBUTING.md
Comment (jirihnidek) [inline CONTRIBUTING.md:104]: When I run this command, then I get following result:

```
echo '"http://localhost:8000/insights_remove.yml"' |   yggctl generate data-message --directive rhc_worker_playbook -
{"type":"data","message_id":"28ab2d71-1077-48fd-bd67-c496cd9a5015","response_to":"","version":1,"sent":"2024-08-12T23:06:28.615688017+02:00","directive":"rhc_worker_playbook","metadata":{},"content":"Imh0dHA6Ly9sb2NhbGhvc3Q6ODAwMC9pbnNpZ2h0c19yZW1vdmUueW1sIgo="}
```

File: CONTRIBUTING.md
Comment (jirihnidek) [inline CONTRIBUTING.md:112]: When I run this command, then I can see following in the log:

```
Aug 12 23:29:38 fedora yggd[14005]: [yggd] 2024/08/12 23:29:38 /builddir/build/BUILD/yggdrasil-0.4.2/_build/src/github.com/redhatinsights/yggdrasil/internal/http/client.go:48: sending HTTP request: GET Imh0dHA6Ly9sb2NhbGhvc3Q6ODAwMC9pbnNpZ2h0c19yZW1vdmUueW1sIgo=
Aug 12 23:29:38 fedora yggd[14005]: [yggd] 2024/08/12 23:29:38 /builddir/build/BUILD/yggdrasil-0.4.2/_build/src/github.com/redhatinsights/yggdrasil/internal/work/dispatcher.go:191: cannot dispatch data: cannot get detached message content: cannot do HTTP request: Get "Imh0dHA6Ly9sb2NhbGhvc3Q6ODAwMC9pbnNpZ2h0c19yZW1vdmUueW1sIgo=": unsupported protocol scheme ""
```

It looks like that there is something wrong with yggdrasil or worker. String "Imh0dHA6Ly9sb2NhbGhvc3Q6ODAwMC9pbnNpZ2h0c19yZW1vdmUueW1sIgo=" is base64 encoded "http://localhost:8000/insights_remove.yml", but it seems that something does not try to decode this string.

File: main.go
Comment (jirihnidek) [inline main.go:85]: The cancel callback function is `nil`. I thought that this worker was the main reason to implement this cancel method. Is it on your TODO list?

**Review by jirihnidek — COMMENTED** [review]
I think that there is some bug in worker or yggdrasil, because it is not possible to call GET method. More details in comments. I also lost few comments, because you pushed some changes during my review. I will continue with testing tomorrow morning.

**Comment by jirihnidek** [comment]
Oh, be sure that the command for generating data message is the new one. You have to use yggdrasil from main branch or some new rpm build of yggdrasil. It is not possible to use yggct from Fedora 40... I had also some old `yggctl` installed in /usr/local/bin.

**Comment by jirihnidek** [comment]
I can send message to worker, but the instruction in `CONTRIBUTING.md` is probably not 100% correct. I can see following message in the log of worker:

```
Aug 13 12:24:19 fedora rhc-worker-playbook[16655]: 2024/08/13 12:24:19 INFO message received message-id=1f6de422-06c8-4527-9e50-3f91ef15478d
Aug 13 12:24:19 fedora rhc-worker-playbook[16655]: 2024/08/13 12:24:19 cannot call rx: invalid metadata: missing return_url
```

BTW: it would be good to add instruction how to see log messages of worker. Something like this: `journalctl -f -b -u com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service`

**Comment by jirihnidek** [comment]
The command should be something like this:

```
echo '"http://localhost:8000/insights_remove.yml"' |   go run ./cmd/yggctl generate data-message --metadata '{"return_url": "http://localhost:8000/", "crc_dispatcher_correlation_id": "foo"}' --directive rhc_worker_playbook - |   pub -broker tcp://localhost:1883 -topic yggdrasil/$(cat /var/lib/yggdrasil/client-id)/data/in
```

It is also necessary to change configuration of worker. It means edit `/etc/rhc-worker-playbook/rhc-worker-playbook.toml`. It should look like this:

```
# rhc-worker-playbook configuration

# recipient directive to register with dispatcher
directive = "rhc_worker_playbook"

# whether to verify incoming playboks
verify-playbook = false

# perform the insights-client GPG check on the insights-core egg
insights-core-gpg-check = false

# how verbose the output should be
log-level = "debug"
```

File: worker.go
Comment (jirihnidek) [inline worker.go:36]: When we do not verify playbook, then it is not necessary to check presence of `crc_dispatcher_correlation_id` in metadata. We could set `correlationId` to empty string in this case.

File: CONTRIBUTING.md
Comment (jirihnidek) [inline CONTRIBUTING.md:62]: I would not recommend to start this server with `nohup` and redirecting everything to `/dev/null` I could be useful to see result of this simple http server.

File: internal/ansible/ansible.go
Comment (jirihnidek) [inline internal/ansible/ansible.go:95]: This log message is not correct, because we actually watch `filepath.Join(privateDataDir, "artifacts", id, "job_events")`. This could be confusing.

File: internal/ansible/ansible.go
Comment (jirihnidek) [inline internal/ansible/ansible.go:122]: Is this go routine actually terminated, when related process is finished? The `started` closure is terminated, but I'm not sure about this go routine.

File: internal/ansible/ansible.go
Comment (jirihnidek) [inline internal/ansible/ansible.go:131]: Does generated error message contain path to the file?

File: internal/ansible/ansible.go
Comment (jirihnidek) [inline internal/ansible/ansible.go:169]: Same comment to this go routine. Is this go routine terminated, when related process is finished?

File: worker.go
Comment (jirihnidek) [inline worker.go:141]: I would add here some comments, why it is necessary to have custom un-marshalers.

File: worker.go
Comment (jirihnidek) [inline worker.go:150]: `playbooks`?

**Review by jirihnidek — COMMENTED** [review]
I have more comments and questions.

File: internal/ansible/ansible.go
Comment (jirihnidek) [reply internal/ansible/ansible.go:169]: My suspicion was correct. This go routine is never terminated despite related process is already finished. You  can test it, when you run trigger some ansible job and after finishing the job you can put something to the status file. e.g.:

```
echo "foo" > /var/local/lib/rhc-worker-playbook/dd2292fb-53d1-4c36-88d6-0ab2a223ec68/artifacts/dd2292fb-53d1-4c36-88d6-0ab2a223ec68/status
```

You will see in the log of worker following message:

```
Aug 15 10:55:37 fedora rhc-worker-playbook[22108]: 2024/08/15 10:55:37 notify event: event=notify.InCloseWrite path=/var/local/lib/rhc-worker-playbook/dd2292fb-53d1-4c36-88d6-0ab2a223ec68/artifacts/dd2292fb-53d1-4c36-88d6-0ab2a223ec68/status
Aug 15 10:55:37 fedora rhc-worker-playbook[22108]: 2024/08/15 10:55:37 unsupported status case: foo
```

File: internal/ansible/ansible.go
Comment (jirihnidek) [reply internal/ansible/ansible.go:122]: My suspicion for this case was also correct. When you create some `foo.json` file with some foo content. Something like this:

```json
{"foo": "bar"}
```

And you move this file to `job_events` directory of finished job:

```
mv ~/foo.json /var/local/lib/rhc-worker-playbook/dd2292fb-53d1-4c36-88d6-0ab2a223ec68/artifacts/dd2292fb-53d1-4c36-88d6-0ab2a223ec68/job_events/foo.json
```

Then it is obvious that the go routine is still running, but it terminates whole worker as you can see:

```
Aug 15 11:27:20 fedora rhc-worker-playbook[22108]: 2024/08/15 11:27:20 notify event: event=notify.InMovedTo path=/var/local/lib/rhc-worker-playbook/dd2292fb-53d1-4c36-88d6-0ab2a223ec68/artifacts/dd2292fb-53d1-4c36-88d6-0ab2a223ec68/job_events/foo.json
Aug 15 11:27:20 fedora rhc-worker-playbook[22108]: 2024/08/15 11:27:20 >>> terminating goroutine for watching data dir: dir=/var/local/lib/rhc-worker-playbook/dd2292fb-53d1-4c36-88d6-0ab2a223ec68
Aug 15 11:27:20 fedora rhc-worker-playbook[22108]: panic: send on closed channel
Aug 15 11:27:20 fedora rhc-worker-playbook[22108]: goroutine 14 [running]:
Aug 15 11:27:20 fedora rhc-worker-playbook[22108]: github.com/redhatinsights/rhc-worker-playbook/internal/ansible.RunPlaybook.func2.1()
Aug 15 11:27:20 fedora rhc-worker-playbook[22108]:         /home/jhnidek/github/redhat-insights/rhc-worker-playbook/internal/ansible/ansible.go:119 +0x60b
Aug 15 11:27:20 fedora rhc-worker-playbook[22108]: created by github.com/redhatinsights/rhc-worker-playbook/internal/ansible.RunPlaybook.func2 in goroutine 11
Aug 15 11:27:20 fedora rhc-worker-playbook[22108]:         /home/jhnidek/github/redhat-insights/rhc-worker-playbook/internal/ansible/ansible.go:93 +0x1bb
Aug 15 11:27:20 fedora systemd[1]: com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Aug 15 11:27:20 fedora systemd[1]: com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service: Failed with result 'exit-code'.
Aug 15 11:27:20 fedora systemd[1]: com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.service: Consumed 1.934s CPU time.
```

... Yeah, I added some extra debug print to the code.

File: internal/ansible/ansible.go
Comment (jirihnidek) [inline internal/ansible/ansible.go:98]: I think that this issue could be solved by creating channel `c` before this go routine and calling this go routine with this channel as argument of the go routine. Stopping and closing channel will terminate the go routine. Suggested change:

```go
		jobEventsChan := make(chan notify.EventInfo, 1)
		if err := notify.Watch(filepath.Join(privateDataDir, "artifacts", id, "job_events"), jobEventsChan, notify.InMovedTo); err != nil {
			log.Errorf("failed to watch private data dir: dir=%v err=%v", privateDataDir, err)
			return
		}
		defer notify.Stop(jobEventsChan)
		defer close(jobEventsChan)
		// start a goroutine to watch for event files being written to the
		// job_events directory. When a relevant event file is detected, it gets
		// marshaled into JSON and sent to the events channel.
		go func(c chan notify.EventInfo) {
			log.Debugf("watching private data dir: dir=%v", privateDataDir)
			defer log.Debugf("stop watching private data dir: dir=%v", privateDataDir)
			for e := range c {
				log.Debugf("notify event: event=%v path=%v", e.Event(), e.Path())
				switch e.Event() {
				case notify.InMovedTo:
					if strings.HasSuffix(e.Path(), ".json") {
						data, err := os.ReadFile(e.Path())
						if err != nil {
							log.Errorf("cannot read file: file=%v error=%v", e.Path(), err)
							continue
						}

						var event Event
						if err := json.Unmarshal(data, &event); err != nil {
							log.Errorf("cannot unmarshal data: data=%v error=%v", data, err)
							continue
						}

						events <- event
						log.Infof("event sent: event=%v", event)
					}
				}
			}
		}(jobEventsChan)
```

**Review by jirihnidek — CHANGES_REQUESTED** [review]
I have another suggestion for solving the issue with watchers.

**Comment by jirihnidek** [comment]
I have also concern about security risks related to access permissions of `/var/lib/rhc-worker-playbook`:

```
[root@fedora rhc-worker-playbook]# ll -d /var/lib/rhc-worker-playbook/
drwxrwx---. 3 root yggdrasil-worker 50 Aug 13 12:44 /var/lib/rhc-worker-playbook/
```

As you can see any worker (using system user that is member of `yggdrasil-worker` group) could be able to read content of this directory, which is not good, because there could be sensitive information about changes of system. I am sure that we do not want to expose such information to other workers.

File: internal/ansible/ansible.go
Comment (subpop) [reply internal/ansible/ansible.go:98]: Excellent idea.

File: data/dbus/com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.conf.in
Comment (subpop) [reply data/dbus/com.redhat.Yggdrasil1.Worker1.rhc_worker_playbook.conf.in:5]: If your findings contradict my understanding here, please let me know.

File: CONTRIBUTING.md
Comment (subpop) [reply CONTRIBUTING.md:112]:
```console
link@lima-fedora:.../Projects/rhc-worker-playbook$ yggctl -v
yggctl version 0.4.2
link@lima-fedora:.../Projects/rhc-worker-playbook$ echo '"http://localhost:8000/insights_remove.yml"' |   yggctl generate data-message --directive rhc_worker_playbook -
{"type":"data","message_id":"0fc2943d-9c1c-4d6e-b3a1-3442d17dd128","response_to":"","version":1,"sent":"2024-08-16T13:08:02.161028419-04:00","directive":"rhc_worker_playbook","metadata":{},"content":"http://localhost:8000/insights_remove.yml"}
```

File: CONTRIBUTING.md
Comment (subpop) [reply CONTRIBUTING.md:104]: This is caused by a bug in older versions of yggctl. Make sure you're using
0.4.2 or later.

**Comment by subpop** [comment]
I'm trying something new in this PR. Instead of rebasing the one big commit with updates from the review feedback, I've pushed a sequence of commits that address the feedback. Once the review is approved, I want to squash all these follow-up commits into the main one.

**Comment by subpop** [comment]
> I also noticed that there is RPM `python3-ansible-runner` available for Fedora 40. Could we use this RPM for upstream version in the future? It is not the best approach to include binary blobs in git repository and ship it bundled in the rhc-worker-playbook RPM. It requires tracking security issues in `ansible-runner`.

Is that RPM going to be carried downstream into CentOS Stream and RHEL? Do we want to become the potential maintainers of it in RHEL for the foreseeable future?

**Comment by subpop** [comment]
> I have also concern about security risks related to access permissions of `/var/lib/rhc-worker-playbook`:
> 
> ```
> [root@fedora rhc-worker-playbook]# ll -d /var/lib/rhc-worker-playbook/
> drwxrwx---. 3 root yggdrasil-worker 50 Aug 13 12:44 /var/lib/rhc-worker-playbook/
> ```
> 
> As you can see any worker (using system user that is member of `yggdrasil-worker` group) could be able to read content of this directory, which is not good, because there could be sensitive information about changes of system. I am sure that we do not want to expose such information to other workers.

Fair point. This is a decision made by the RPM rather than the client itself, but tightening the permissions on the directory can't hurt.

File: internal/ansible/ansible.go
Comment (jirihnidek) [reply internal/ansible/ansible.go:137]: I can confirm that without this proposed change the goroutine is never finished and following message `"stop goroutine watching status file: ..."` is not printed to log.

**Review by jirihnidek — CHANGES_REQUESTED** [review]
Thanks for all updates. I have only one final request.

File: internal/ansible/ansible.go
Comment (subpop) [reply internal/ansible/ansible.go:137]: Oops. Oversight on my part. Thanks for catching it.

**Review by jirihnidek — APPROVED** [review]
LGTM, you can squash commits as you proposed.

**Pull request merged**
